src/main/flight/gps_rescue.c

```diff
@@ -417,24 +291,204 @@ void rescueAttainPosition()
     DEBUG_SET(DEBUG_RTH, 3, rescueState.failure);
 }
 
-// Very similar to maghold function on betaflight/cleanflight
-void setBearing(int16_t desiredHeading)
+static void performSanityChecks()
 {
-    float errorAngle = (attitude.values.yaw / 10.0f) - desiredHeading;
+    if (rescueState.phase == RESCUE_IDLE) {
+        rescueState.failure = RESCUE_HEALTHY;
 
-    // Determine the most efficient direction to rotate
-    if (errorAngle <= -180) {
-        errorAngle += 360;
-    } else if (errorAngle > 180) {
-        errorAngle -= 360;
+        return;
     }
 
-    errorAngle *= -GET_DIRECTION(rcControlsConfig()->yaw_control_reversed);
+    // Do not abort until each of these items is fully tested
+    if (rescueState.failure != RESCUE_HEALTHY) {
+        if (gpsRescueConfig()->sanityChecks == RESCUE_SANITY_ON
+            || (gpsRescueConfig()->sanityChecks == RESCUE_SANITY_FS_ONLY && rescueState.isFailsafe == true)) {
+            rescueState.phase = RESCUE_ABORT;
+        }
+    }
 
-    // Calculate a desired yaw rate based on a maximum limit beyond
-    // an error window and then scale the requested rate down inside
-    // the window as error approaches 0.
-    rescueYaw = -constrainf(errorAngle / GPS_RESCUE_RATE_SCALE_DEGREES * GPS_RESCUE_MAX_YAW_RATE, -GPS_RESCUE_MAX_YAW_RATE, GPS_RESCUE_MAX_YAW_RATE);
+    // Check if crash recovery mode is active, disarm if so.
+    if (crashRecoveryModeActive()) {
+        rescueState.failure = RESCUE_CRASH_DETECTED;
+    }
+
+    //  Things that should run at a low refresh rate (such as flyaway detection, etc)
+    //  This runs at ~1hz
+
+    static uint32_t previousTimeUs;
+
+    const uint32_t currentTimeUs = micros();
+    const uint32_t dTime = currentTimeUs - previousTimeUs;
+
+    if (dTime < 1000000) { //1hz
+        return;
+    }
+
+    previousTimeUs = currentTimeUs;
+
+     // Stalled movement detection
+    static int8_t gsI = 0;
+
+    gsI = constrain(gsI + (rescueState.sensor.groundSpeed < 150) ? 1 : -1, -10, 10);
+
+    if (gsI == 10) {
+        rescueState.failure = RESCUE_CRASH_DETECTED;
+    }
+
+    // Minimum sat detection
+    static int8_t msI = 0;
+
+    msI = constrain(msI + (rescueState.sensor.numSat < gpsRescueConfig()->minSats) ? 1 : -1, -5, 5);
+
+    if (msI == 5) {
+        rescueState.failure = RESCUE_FLYAWAY;
+    }
+}
+
+static void sensorUpdate()
+{
+    rescueState.sensor.currentAltitudeCm = getEstimatedAltitudeCm();
+
+    // Calculate altitude velocity
+    static uint32_t previousTimeUs;
+    static int32_t previousAltitudeCm;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The scope of the variable 'previousAltitudeCm' can be reduced.](https://app.codacy.com/gh/emuflight/EmuFlight/pullRequest?prid=5552604)